## **Pandas Essencial**
**Prof. Dr. Samuel Martins (@hisamuka @xavecoding)** <br/>
xavecoding: https://youtube.com/c/xavecoding <br/>

Neste tutorial, vamos aprender o essencial da biblioteca _Pandas_ para manipulação de dados.<br/><br/>

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

### Pacotes usados neste Notebook

In [2]:
# pacotes usados neste notebook
import pandas as pd

<h1>Manipulação de Dados com Pandas</h1><hr/>

<h2>1. Manipulação Básica de Datasets</h2>
<hr/>

**Dataset**: Gas Prices in Brazil: https://www.kaggle.com/matheusfreitag/gas-prices-in-brazil <br/>

Este dataset contém os **registros dos preços médios semanais dos combustíveis do Brasil entre os anos de 2004 e 2019**. <br/>
Cada *Registro (registro/linha)* consiste em um registro de preço aferido para um dado tipo de combustível em uma dada localidade do Brasil. <br/>
Alguns dos principais *atributos* (colunas) do dataset são: 'ESTADO', 'PRODUTO', 'NÚMERO DE POSTOS PESQUISADOS', 'PREÇO MÉDIO REVENDA'.


\* O arquivo disponibilizado no Kaggle está no formato *tsv*. Embora o _pandas_ consiga abrí-lo normalmente, convertemos tal arquivo para o formato *CSV*, que é um dos formatos mais utilizados, e mudamos seu separado para ';' apenas para mostrar algumas opções da função de carregamento.

### 1.1. Importando o Dataset
Para carregar um dataset no formato csv, basta utilizar a função `read_csv` do pandas. Por padrão, ela considera _','_ como separador.

In [3]:
# chamamos o pandas e indicamos a leitura de um csv, o sep está indicando o separador de leitura
data = pd.read_csv('./datasets/GasPricesinBrazil_2004-2019.csv', sep=',')

In [4]:
data

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106818               GNV                             7             R$/m3   
106819               GNV                            23             R$/m3   
106820               GNV                            24             R$/m3   
106821               GNV                            52             R$/m3   
106822               GNV                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                     1.288                  0.016                 1.190   
1                     1.162                  0.114                 0.890   
2                     1.389                  0.097                 1.180   
3                     1.262                  0.070                 1.090   
4                     1.181                  0.078                 1.050   
...                     ...                    ...                   ...   
106818                3.574                  0.065                 3.499   
106819                3.401                  0.129                 3.230   
106820                2.912                  0.190                 2.699   
106821                3.020                  0.229                 2.699   
106822                3.697                  0.005                 3.689   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                      1.350                0.463                     0.012   
1                      1.449                0.399                     0.098   
2                      1.760                0.419                     0.070   
3                      1.509                0.432                     0.055   
4                      1.400                 0.24                     0.066   
...                      ...                  ...                       ...   
106818                 3.690                0.818                     0.018   
106819                 3.789                 0.95                     0.038   
106820                 3.499                0.914                     0.065   
106821                 3.490                0.646                     0.076   
106822                 3.699                1.179                     0.001   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                         0.825                       0.11   
1                         0.763                      0.088   
2                          0.97                      0.095   
3                          0.83                      0.119   
4                         0.941                      0.077   
...                

O dataset não foi carregado corretamente pois o separador utilizado seu arquivo CSV era ';' e não a ','. <br/>
Vamos então carregá-lo corretamente:

### 1.2. Exibindo as primeiras linhas do Dataset
A função `.head()` exibe as 5 primeiras linhas do dataset/tabela/Data Frame.

In [69]:
#por padrao exibe 5, se passarmos um parametro, ira corresponder a ele. 
data.head(10)

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   
5   2004-05-09  2004-05-15      NORDESTE               BAHIA   
6   2004-05-09  2004-05-15      NORDESTE               CEARA   
7   2004-05-09  2004-05-15      NORDESTE            MARANHAO   
8   2004-05-09  2004-05-15      NORDESTE             PARAIBA   
9   2004-05-09  2004-05-15      NORDESTE          PERNAMBUCO   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   
5  ETANOL HIDRATADO                           408              R$/l   
6  ETANOL HIDRATADO                           278              R$/l   
7  ETANOL HIDRATADO                           105              R$/l   
8  ETANOL HIDRATADO                           125              R$/l   
9  ETANOL HIDRATADO                           423              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                 1.190   
1                1.162                  0.114                 0.890   
2                1.389                  0.097                 1.180   
3                1.262                  0.070                 1.090   
4                1.181                  0.078                 1.050   
5                1.383                  0.132                 0.999   
6                1.453                  0.218                 1.030   
7                1.631                  0.158                 1.350   
8                1.284                  0.130                 1.100   
9                1.224                  0.141                 0.989   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   
3                 1.509                0.432                     0.055   
4                 1.400                 0.24                     0.066   
5                 2.050                0.426                     0.095   
6                 1.950                0.353                     0.150   
7                 2.000                0.515                     0.097   
8                 1.699                0.353                     0.101   
9                 1.700                0.277                     0.115   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   
3                     0.83                      0.119   
4                    0.941                      0.077   
5                    0.957                      0.128   
6                      1.1                      0.068   
7                    1.116                      0.091   
8                    0.931                      0.072   
9                    0.947                      0.091   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   
3                    0.5991                   1.22242   
4    

### 1.3 Informações do Dataset e Elementos Chave

### 1.3.1 Informações gerais sobre o Dataset

In [70]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106823 entries, 0 to 106822
Data columns (total 20 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   DATA INICIAL                   106823 non-null  object 
 1   DATA FINAL                     106823 non-null  object 
 2   REGIÃO                         106823 non-null  object 
 3   ESTADO                         106823 non-null  object 
 4   PRODUTO                        106823 non-null  object 
 5   NÚMERO DE POSTOS PESQUISADOS   106823 non-null  int64  
 6   UNIDADE DE MEDIDA              106823 non-null  object 
 7   PREÇO MÉDIO REVENDA            106823 non-null  float64
 8   DESVIO PADRÃO REVENDA          106823 non-null  float64
 9   PREÇO MÍNIMO REVENDA           106823 non-null  float64
 10  PREÇO MÁXIMO REVENDA           106823 non-null  float64
 11  MARGEM MÉDIA REVENDA           106823 non-null  object 
 12  COEF DE VARIAÇÃO REVENDA      

A primeira coluna da tabela, chamada _'Unnamed: 0'_, parece não significar nada. Na verdade, ela parece ser os **índices** da tabela, que foram salvos como uma _coluna_.<br/>
Veremos jajá como **remover tal coluna**.

Outro ponto é que, aparentemente, nenhum atributo/coluna possui valores nulos (_null_), uma vez que o número de registros do dataframe e os números de valores _non-null_ é de **106823**. <br/>
Mas, veremos que não é bem assim para esse caso.

### 1.3.2 Data Frame
Todo dataset carregado (dados estruturados) é um `Data Frame`: 'Tabela' bi-dimensional, de tamanho mutável, com dados potencialmente heterogêneos. <br/>

In [71]:
type(data)

pandas.core.frame.DataFrame

Podemos acessar as **dimensões do Data Frame** (número de linhas x número de colunas) utilizando o atributo `.shape` do Data Frame.

In [72]:
data.shape

(106823, 20)

In [73]:
print(f"O dataframe possui {data.shape[0]} linhas e {data.shape[1]} Colunas")

O dataframe possui 106823 linhas e 20 Colunas


#### **Criando um DataFrame**

Podemos criar um DataFrame a partir de um _dicionário_, onde cada **chave** possui uma **lista de elementos de igual tamanho**.<br/>
As **chaves** representam as **colunas** e **cada um dos valores de sua lista** representa o **valor da linha** correspondente para aquela coluna.

In [74]:
alunos_df = pd.DataFrame({
    'nome': ['Luke Skywalker', 'Yoda', 'Palpatine'],
    'idade': [16, 1000, 70],
    'peso': [70, 15, 60],
    'eh jedi': [True, True, False]  # o nome das colunas podem ter espaços
})

In [75]:
alunos_df

nome  idade  peso  eh jedi
0  Luke Skywalker     16    70     True
1            Yoda   1000    15     True
2       Palpatine     70    60    False

#### **VEJA MAIS**
Criando um Data Frame a partir de um dicionário: https://www.geeksforgeeks.org/how-to-create-dataframe-from-dictionary-in-python-pandas/

#### **Renomeando as colunas de um DataFrame**
**===>** O método `DataFrame.columns` retorna uma ";lista" com os **nomes de todas as colunas** do data frame.

In [76]:
alunos_df.columns

Index(['nome', 'idade', 'peso', 'eh jedi'], dtype='object')

In [77]:
list(alunos_df)

['nome', 'idade', 'peso', 'eh jedi']

<br/>

**===>** Para **renomear colunas** do data frame, utilize o método `DataFrame.rename`, que retorna uma _cópia_ do data frame com as as colunas renomeadas:

In [78]:
alunos_df.rename(columns={
    'nome': 'Nomes',
    'idade': 'Idade',
    'peso': 'Peso'
}, inplace=True)

Para renomear o _próprio_ data frame em questão, utilize o parâmetro `inplace=True`:

In [79]:
alunos = alunos_df.rename(columns={
    'nome': 'Nomes',
    'idade': 'Idade',
    'peso': 'Peso'
})

In [80]:
alunos

Nomes  Idade  Peso  eh jedi
0  Luke Skywalker     16    70     True
1            Yoda   1000    15     True
2       Palpatine     70    60    False

<br/>

**===>** Uma outra forma de **renomear todas as listas** de um data frame é passar uma _lista_ com os novos nomes das colunas para o atributo `DataFrame.columns`:

In [81]:
alunos.columns = [ "NOME", 'IDADE', 'PESO', 'EH_JEDI']

In [82]:
alunos

NOME  IDADE  PESO  EH_JEDI
0  Luke Skywalker     16    70     True
1            Yoda   1000    15     True
2       Palpatine     70    60    False

### 1.3.3 Series

Array uni-dimensional com os dados e rótulos de um eixo.

In [83]:
data.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                  1.19   
1                1.162                  0.114                  0.89   
2                1.389                  0.097                  1.18   
3                1.262                  0.070                  1.09   
4                1.181                  0.078                  1.05   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   
3                 1.509                0.432                     0.055   
4                 1.400                 0.24                     0.066   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   
3                     0.83                      0.119   
4                    0.941                      0.077   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   
3                    0.5991                   1.22242   
4                    0.7441                    1.0317   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  MÊS   ANO  
0                         0.133    5  2004  
1                         0.115    5  2004  
2                         0.098    5  2004  
3                         0.143    5  2004  
4                         0.082    5  2004

In [84]:
#selecionado unma coluna inteira
data['ESTADO']

0            DISTRITO FEDERAL
1                       GOIAS
2                 MATO GROSSO
3          MATO GROSSO DO SUL
4                     ALAGOAS
                 ...         
106818    RIO GRANDE DO NORTE
106819      RIO GRANDE DO SUL
106820         SANTA CATARINA
106821              SAO PAULO
106822                SERGIPE
Name: ESTADO, Length: 106823, dtype: object

In [85]:
#também funciona para listar a SERIES, entretanto nao pode ter espaço nem caracter especial.
data.PRODUTO

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

In [86]:
data.iloc[1]

DATA INICIAL                           2004-05-09
DATA FINAL                             2004-05-15
REGIÃO                               CENTRO OESTE
ESTADO                                      GOIAS
PRODUTO                          ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS                  387
UNIDADE DE MEDIDA                            R$/l
PREÇO MÉDIO REVENDA                         1.162
DESVIO PADRÃO REVENDA                       0.114
PREÇO MÍNIMO REVENDA                         0.89
PREÇO MÁXIMO REVENDA                        1.449
MARGEM MÉDIA REVENDA                        0.399
COEF DE VARIAÇÃO REVENDA                    0.098
PREÇO MÉDIO DISTRIBUIÇÃO                    0.763
DESVIO PADRÃO DISTRIBUIÇÃO                  0.088
PREÇO MÍNIMO DISTRIBUIÇÃO                  0.5013
PREÇO MÁXIMO DISTRIBUIÇÃO                    1.05
COEF DE VARIAÇÃO DISTRIBUIÇÃO               0.115
MÊS                                             5
ANO                                          2004


In [87]:
type(data.iloc[1])

pandas.core.series.Series

#### **Criando uma Series**

Podemos criar um DataFrame a partir de uma lista de elementos.

In [88]:
pd.Series([1.7, 5.5, 6.0, 9.5])

0    1.7
1    5.5
2    6.0
3    9.5
dtype: float64

Podemos alterar o **nome dos índices** (veremos melhor jajá) e o **nome da Series** (o que ela representa):

In [89]:
#atentar-se que o numero de indices tem de ser igual ao total de itens.
pd.Series([1.7, 5.5, 6.0, 9.5], index=['prova 1', 'prova 2', 'projeto', 'prova5'], name='Notas do aluno') 


prova 1    1.7
prova 2    5.5
projeto    6.0
prova5     9.5
Name: Notas do aluno, dtype: float64

In [90]:
data.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                1.288                  0.016                  1.19   
1                1.162                  0.114                  0.89   
2                1.389                  0.097                  1.18   
3                1.262                  0.070                  1.09   
4                1.181                  0.078                  1.05   

   PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                 1.350                0.463                     0.012   
1                 1.449                0.399                     0.098   
2                 1.760                0.419                     0.070   
3                 1.509                0.432                     0.055   
4                 1.400                 0.24                     0.066   

  PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                    0.825                       0.11   
1                    0.763                      0.088   
2                     0.97                      0.095   
3                     0.83                      0.119   
4                    0.941                      0.077   

  PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                    0.4201                    0.9666   
1                    0.5013                      1.05   
2                    0.5614                     1.161   
3                    0.5991                   1.22242   
4                    0.7441                    1.0317   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO  MÊS   ANO  
0                         0.133    5  2004  
1                         0.115    5  2004  
2                         0.098    5  2004  
3                         0.143    5  2004  
4                         0.082    5  2004

#### **VEJA MAIS**
https://pandas.pydata.org/pandas-docs/stable/user_guide/dsintro.html

### 1.3.4 Atribuindo Dados

#### 1.3.4.1 Atribuindo constantes

In [91]:
# a series retornada referente a coluna, nao é uma cópia, mas sim, uma referencia/view á coluna do dataframe
data['PRODUTO']

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

In [92]:
produto_copy_bkp = data['PRODUTO'].copy() # retorna uma cópia da coluna 'PRODUTO'
produto_copy_bkp

0         ETANOL HIDRATADO
1         ETANOL HIDRATADO
2         ETANOL HIDRATADO
3         ETANOL HIDRATADO
4         ETANOL HIDRATADO
                ...       
106818                 GNV
106819                 GNV
106820                 GNV
106821                 GNV
106822                 GNV
Name: PRODUTO, Length: 106823, dtype: object

In [93]:
data['PRODUTO'] = 'Combustivel'
data.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO      PRODUTO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL  Combustivel   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS  Combustivel   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO  Combustivel   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL  Combustivel   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS  Combustivel   

   NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
0                           127              R$/l                1.288   
1                           387              R$/l                1.162   
2                           192              R$/l                1.389   
3                           162              R$/l                1.262   
4                           103              R$/l                1.181   

   DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  PREÇO MÁXIMO REVENDA  \
0                  0.016                  1.19                 1.350   
1                  0.114                  0.89                 1.449   
2                  0.097                  1.18                 1.760   
3                  0.070                  1.09                 1.509   
4                  0.078                  1.05                 1.400   

  MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                0.463                     0.012                    0.825   
1                0.399                     0.098                    0.763   
2                0.419                     0.070                     0.97   
3                0.432                     0.055                     0.83   
4                 0.24                     0.066                    0.941   

  DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                       0.11                    0.4201   
1                      0.088                    0.5013   
2                      0.095                    0.5614   
3                      0.119                    0.5991   
4                      0.077                    0.7441   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO  MÊS   ANO  
0                    0.9666                         0.133    5  2004  
1                      1.05                         0.115    5  2004  
2                     1.161                         0.098    5  2004  
3                   1.22242                         0.143    5  2004  
4                    1.0317                         0.082    5  2004

In [94]:
nrows, ncols = data.shape
nrows, ncols

(106823, 20)

In [95]:
novos_produtos = [f'Produto  {i}' for i in range(nrows)] #passamos 'produto' em seu indice
len(novos_produtos)

106823

In [96]:
data['PRODUTO'] = novos_produtos #essa atribuição deu certo por ter o mesmo tamanho de linhas

In [97]:
data

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0            Produto  0                           127              R$/l   
1            Produto  1                           387              R$/l   
2            Produto  2                           192              R$/l   
3            Produto  3                           162              R$/l   
4            Produto  4                           103              R$/l   
...                 ...                           ...               ...   
106818  Produto  106818                             7             R$/m3   
106819  Produto  106819                            23             R$/m3   
106820  Produto  106820                            24             R$/m3   
106821  Produto  106821                            52             R$/m3   
106822  Produto  106822                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                     1.288                  0.016                 1.190   
1                     1.162                  0.114                 0.890   
2                     1.389                  0.097                 1.180   
3                     1.262                  0.070                 1.090   
4                     1.181                  0.078                 1.050   
...                     ...                    ...                   ...   
106818                3.574                  0.065                 3.499   
106819                3.401                  0.129                 3.230   
106820                2.912                  0.190                 2.699   
106821                3.020                  0.229                 2.699   
106822                3.697                  0.005                 3.689   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                      1.350                0.463                     0.012   
1                      1.449                0.399                     0.098   
2                      1.760                0.419                     0.070   
3                      1.509                0.432                     0.055   
4                      1.400                 0.24                     0.066   
...                      ...                  ...                       ...   
106818                 3.690                0.818                     0.018   
106819                 3.789                 0.95                     0.038   
106820                 3.499                0.914                     0.065   
106821                 3.490                0.646                     0.076   
106822                 3.699                1.179                     0.001   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                         0.825                       0.11   
1                         0.763                      0.088   
2                          0.97                      0.095   
3                          0.83                      0.119   
4                         0.941                      0.077   
...                         ...

#### 1.3.4.2 Atribuindo listas ou series

In [98]:
data['PRODUTO'] = produto_copy_bkp #estamos retornando os valores de produto para seus dados originais
data

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106818               GNV                             7             R$/m3   
106819               GNV                            23             R$/m3   
106820               GNV                            24             R$/m3   
106821               GNV                            52             R$/m3   
106822               GNV                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  \
0                     1.288                  0.016                 1.190   
1                     1.162                  0.114                 0.890   
2                     1.389                  0.097                 1.180   
3                     1.262                  0.070                 1.090   
4                     1.181                  0.078                 1.050   
...                     ...                    ...                   ...   
106818                3.574                  0.065                 3.499   
106819                3.401                  0.129                 3.230   
106820                2.912                  0.190                 2.699   
106821                3.020                  0.229                 2.699   
106822                3.697                  0.005                 3.689   

        PREÇO MÁXIMO REVENDA MARGEM MÉDIA REVENDA  COEF DE VARIAÇÃO REVENDA  \
0                      1.350                0.463                     0.012   
1                      1.449                0.399                     0.098   
2                      1.760                0.419                     0.070   
3                      1.509                0.432                     0.055   
4                      1.400                 0.24                     0.066   
...                      ...                  ...                       ...   
106818                 3.690                0.818                     0.018   
106819                 3.789                 0.95                     0.038   
106820                 3.499                0.914                     0.065   
106821                 3.490                0.646                     0.076   
106822                 3.699                1.179                     0.001   

       PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                         0.825                       0.11   
1                         0.763                      0.088   
2                          0.97                      0.095   
3                          0.83                      0.119   
4                         0.941                      0.077   
...                

#### 1.3.4.3 Criando novas colunas
Para **criar uma nova coluna** em um data frame, basta atribuirmos uma _lista/Series de valores_ a uma **nova 'chave'** do data frame. <br/>

**PS:** A _quantidade de valores_ da lista precisa ser **igual** ao _número de linhas/registros_ do data frame.

In [99]:
#estamos criando uma nova coluna com os valores declarados como default
data['coluna sem nocao'] = 'DEFAULT'
data 

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106818               GNV                             7             R$/m3   
106819               GNV                            23             R$/m3   
106820               GNV                            24             R$/m3   
106821               GNV                            52             R$/m3   
106822               GNV                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                     1.288                  0.016                 1.190  ...   
1                     1.162                  0.114                 0.890  ...   
2                     1.389                  0.097                 1.180  ...   
3                     1.262                  0.070                 1.090  ...   
4                     1.181                  0.078                 1.050  ...   
...                     ...                    ...                   ...  ...   
106818                3.574                  0.065                 3.499  ...   
106819                3.401                  0.129                 3.230  ...   
106820                2.912                  0.190                 2.699  ...   
106821                3.020                  0.229                 2.699  ...   
106822                3.697                  0.005                 3.689  ...   

        MARGEM MÉDIA REVENDA COEF DE VARIAÇÃO REVENDA  \
0                      0.463                    0.012   
1                      0.399                    0.098   
2                      0.419                    0.070   
3                      0.432                    0.055   
4                       0.24                    0.066   
...                      ...                      ...   
106818                 0.818                    0.018   
106819                  0.95                    0.038   
106820                 0.914                    0.065   
106821                 0.646                    0.076   
106822                 1.179                    0.001   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                          0.825                       0.11   
1                          0.763                      0.088   
2                           0.97                      0.095   
3                           0.83                      0.119   
4                          0.941                      0.077   
...                          ...                        ...   
106818                     2.756                          0   
106819                     2.451                      0.402   
106820                     1

In [100]:
#criamos uma coluna nova com o valor sendo o numero de linhas
data['coluna a partir de lista'] = range(data.shape[0]) 
data

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106818               GNV                             7             R$/m3   
106819               GNV                            23             R$/m3   
106820               GNV                            24             R$/m3   
106821               GNV                            52             R$/m3   
106822               GNV                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                     1.288                  0.016                 1.190  ...   
1                     1.162                  0.114                 0.890  ...   
2                     1.389                  0.097                 1.180  ...   
3                     1.262                  0.070                 1.090  ...   
4                     1.181                  0.078                 1.050  ...   
...                     ...                    ...                   ...  ...   
106818                3.574                  0.065                 3.499  ...   
106819                3.401                  0.129                 3.230  ...   
106820                2.912                  0.190                 2.699  ...   
106821                3.020                  0.229                 2.699  ...   
106822                3.697                  0.005                 3.689  ...   

        COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                          0.012                    0.825   
1                          0.098                    0.763   
2                          0.070                     0.97   
3                          0.055                     0.83   
4                          0.066                    0.941   
...                          ...                      ...   
106818                     0.018                    2.756   
106819                     0.038                    2.451   
106820                     0.065                    1.998   
106821                     0.076                    2.374   
106822                     0.001                    2.518   

        DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                             0.11                    0.4201   
1                            0.088                    0.5013   
2                            0.095                    0.5614   
3                            0.119                    0.5991   
4                            0.077                    0.7441   
...                            ...                       ...   
106818                           0                     2.756   
106819                       0.402 

<br/>

Outro exemplo:

In [101]:
data.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                1.288                  0.016                  1.19  ...   
1                1.162                  0.114                  0.89  ...   
2                1.389                  0.097                  1.18  ...   
3                1.262                  0.070                  1.09  ...   
4                1.181                  0.078                  1.05  ...   

   COEF DE VARIAÇÃO REVENDA PREÇO MÉDIO DISTRIBUIÇÃO  \
0                     0.012                    0.825   
1                     0.098                    0.763   
2                     0.070                     0.97   
3                     0.055                     0.83   
4                     0.066                    0.941   

   DESVIO PADRÃO DISTRIBUIÇÃO PREÇO MÍNIMO DISTRIBUIÇÃO  \
0                        0.11                    0.4201   
1                       0.088                    0.5013   
2                       0.095                    0.5614   
3                       0.119                    0.5991   
4                       0.077                    0.7441   

  PREÇO MÁXIMO DISTRIBUIÇÃO COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO  \
0                    0.9666                         0.133   5  2004   
1                      1.05                         0.115   5  2004   
2                     1.161                         0.098   5  2004   
3                   1.22242                         0.143   5  2004   
4                    1.0317                         0.082   5  2004   

   coluna sem nocao  coluna a partir de lista  
0           DEFAULT                         0  
1           DEFAULT                         1  
2           DEFAULT                         2  
3           DEFAULT                         3  
4           DEFAULT                         4  

[5 rows x 22 columns]

In [102]:
data['PREÇO MÉDIO REVENDA (dólares)'] = data['PREÇO MÉDIO REVENDA'] *5.25
data

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106818               GNV                             7             R$/m3   
106819               GNV                            23             R$/m3   
106820               GNV                            24             R$/m3   
106821               GNV                            52             R$/m3   
106822               GNV                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                     1.288                  0.016                 1.190  ...   
1                     1.162                  0.114                 0.890  ...   
2                     1.389                  0.097                 1.180  ...   
3                     1.262                  0.070                 1.090  ...   
4                     1.181                  0.078                 1.050  ...   
...                     ...                    ...                   ...  ...   
106818                3.574                  0.065                 3.499  ...   
106819                3.401                  0.129                 3.230  ...   
106820                2.912                  0.190                 2.699  ...   
106821                3.020                  0.229                 2.699  ...   
106822                3.697                  0.005                 3.689  ...   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                          0.825                       0.11   
1                          0.763                      0.088   
2                           0.97                      0.095   
3                           0.83                      0.119   
4                          0.941                      0.077   
...                          ...                        ...   
106818                     2.756                          0   
106819                     2.451                      0.402   
106820                     1.998                          0   
106821                     2.374                      0.165   
106822                     2.518                          0   

        PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                          0.4201                    0.9666   
1                          0.5013                      1.05   
2                          0.5614                     1.161   
3                          0.5991                   1.22242   
4                          0.7441                    1.0317   
...                           ...                       ...   
106818                      2.756                     2.756   
106819             

**PS:** Obviamente, a lógica correta em converter o preço dos combustíveis em reais para dólares não é considerar uma taxa de câmbio fixa, uma vez que cada preço foi aferido em um momento diferente.

### 1.3.4 Índices

Todo Data Frame possui **índices**, que não são considerado colunas da tabela. Tais índices são comumente **númericos**, de 0 a num_linhas-1, mas também podem ser **textuais (rótulos/labels)**.

In [103]:
data.index

RangeIndex(start=0, stop=106823, step=1)

In [104]:
list(data.index.to_list())

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


Use `list(data.index)` ou `data.index.to_list()` para converter um RangeIndex para uma python list.

#### **Exemplo de Data Frame com Índices Textuais (labels)**

In [105]:
pesquisa_de_satisfacao = pd.DataFrame({
    'bom': [50, 21, 100],
    'ruim': [131, 2, 30],
    'pessimo': [30, 20, 1]
}, index=['XboxOne', 'Playstation4', 'Switch'])

In [106]:
pesquisa_de_satisfacao.head()

bom  ruim  pessimo
XboxOne        50   131       30
Playstation4   21     2       20
Switch        100    30        1

### 1.4 Selecionando uma ou mais amostras (Indexação)

#### **==>  Index-based selection (seleção baseada em Índices)**
Mostrando linhas específicas de um DataFrame:

`iloc`: seleciona elementos do Dataframe, baseado em seu **índice (número)** --> row-first, column-second

**Selecionando uma amostra/linha:**

In [107]:
data.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                1.288                  0.016                  1.19  ...   
1                1.162                  0.114                  0.89  ...   
2                1.389                  0.097                  1.18  ...   
3                1.262                  0.070                  1.09  ...   
4                1.181                  0.078                  1.05  ...   

   PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                     0.825                       0.11   
1                     0.763                      0.088   
2                      0.97                      0.095   
3                      0.83                      0.119   
4                     0.941                      0.077   

   PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                     0.4201                    0.9666   
1                     0.5013                      1.05   
2                     0.5614                     1.161   
3                     0.5991                   1.22242   
4                     0.7441                    1.0317   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO coluna sem nocao  \
0                         0.133   5  2004          DEFAULT   
1                         0.115   5  2004          DEFAULT   
2                         0.098   5  2004          DEFAULT   
3                         0.143   5  2004          DEFAULT   
4                         0.082   5  2004          DEFAULT   

   coluna a partir de lista  PREÇO MÉDIO REVENDA (dólares)  
0                         0                        6.76200  
1                         1                        6.10050  
2                         2                        7.29225  
3                         3                        6.62550  
4                         4                        6.20025  

[5 rows x 23 columns]

In [108]:
# selecionando a linha 1 ===> observação de indice [1] do dataframe
data.iloc[1]

DATA INICIAL                           2004-05-09
DATA FINAL                             2004-05-15
REGIÃO                               CENTRO OESTE
ESTADO                                      GOIAS
PRODUTO                          ETANOL HIDRATADO
NÚMERO DE POSTOS PESQUISADOS                  387
UNIDADE DE MEDIDA                            R$/l
PREÇO MÉDIO REVENDA                         1.162
DESVIO PADRÃO REVENDA                       0.114
PREÇO MÍNIMO REVENDA                         0.89
PREÇO MÁXIMO REVENDA                        1.449
MARGEM MÉDIA REVENDA                        0.399
COEF DE VARIAÇÃO REVENDA                    0.098
PREÇO MÉDIO DISTRIBUIÇÃO                    0.763
DESVIO PADRÃO DISTRIBUIÇÃO                  0.088
PREÇO MÍNIMO DISTRIBUIÇÃO                  0.5013
PREÇO MÁXIMO DISTRIBUIÇÃO                    1.05
COEF DE VARIAÇÃO DISTRIBUIÇÃO               0.115
MÊS                                             5
ANO                                          2004


**Selecionando múltiplias amostras/linhas:**

In [109]:
#selecinando as linhas de indice de 0 a 5 (incluso)
data.iloc[:6]

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   
5   2004-05-09  2004-05-15      NORDESTE               BAHIA   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   
5  ETANOL HIDRATADO                           408              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                1.288                  0.016                 1.190  ...   
1                1.162                  0.114                 0.890  ...   
2                1.389                  0.097                 1.180  ...   
3                1.262                  0.070                 1.090  ...   
4                1.181                  0.078                 1.050  ...   
5                1.383                  0.132                 0.999  ...   

   PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                     0.825                       0.11   
1                     0.763                      0.088   
2                      0.97                      0.095   
3                      0.83                      0.119   
4                     0.941                      0.077   
5                     0.957                      0.128   

   PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                     0.4201                    0.9666   
1                     0.5013                      1.05   
2                     0.5614                     1.161   
3                     0.5991                   1.22242   
4                     0.7441                    1.0317   
5                     0.5686                      1.35   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO coluna sem nocao  \
0                         0.133   5  2004          DEFAULT   
1                         0.115   5  2004          DEFAULT   
2                         0.098   5  2004          DEFAULT   
3                         0.143   5  2004          DEFAULT   
4                         0.082   5  2004          DEFAULT   
5                         0.134   5  2004          DEFAULT   

   coluna a partir de lista  PREÇO MÉDIO REVENDA (dólares)  
0                         0                        6.76200  
1                         1                        6.10050  
2                         2                        7.29225  
3                         3                        6.62550  
4                         4                        6.20025  
5                         5                        7.26075  

[6 rows x 23 columns]

In [110]:
#selecionando asmlinhas de indice de 10 a 15 (incluso)
data.iloc[10:16]

DATA INICIAL  DATA FINAL    REGIÃO               ESTADO           PRODUTO  \
10   2004-05-09  2004-05-15  NORDESTE                PIAUI  ETANOL HIDRATADO   
11   2004-05-09  2004-05-15  NORDESTE  RIO GRANDE DO NORTE  ETANOL HIDRATADO   
12   2004-05-09  2004-05-15  NORDESTE              SERGIPE  ETANOL HIDRATADO   
13   2004-05-09  2004-05-15     NORTE                 ACRE  ETANOL HIDRATADO   
14   2004-05-09  2004-05-15     NORTE                AMAPA  ETANOL HIDRATADO   
15   2004-05-09  2004-05-15     NORTE             AMAZONAS  ETANOL HIDRATADO   

    NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
10                            89              R$/l                1.573   
11                           131              R$/l                1.276   
12                            57              R$/l                1.301   
13                            33              R$/l                1.805   
14                            15              R$/l                1.785   
15                            53              R$/l                1.578   

    DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
10                  0.134                  1.40  ...   
11                  0.116                  1.09  ...   
12                  0.109                  1.15  ...   
13                  0.134                  1.50  ...   
14                  0.035                  1.74  ...   
15                  0.123                  1.28  ...   

    PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
10                     1.079                      0.078   
11                     0.944                      0.078   
12                      0.98                      0.052   
13                     1.116                      0.171   
14                     1.373                      0.167   
15                     1.137                      0.068   

    PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
10                     0.8857                    1.2499   
11                     0.7119                    1.1386   
12                     0.8267                    1.1059   
13                      0.767                    1.2992   
14                     0.8339                    1.4912   
15                     0.9515                    1.2109   

   COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO coluna sem nocao  \
10                         0.072   5  2004          DEFAULT   
11                         0.083   5  2004          DEFAULT   
12                         0.053   5  2004          DEFAULT   
13                         0.153   5  2004          DEFAULT   
14                         0.122   5  2004          DEFAULT   
15                          0.06   5  2004          DEFAULT   

    coluna a partir de lista  PREÇO MÉDIO REVENDA (dólares)  
10                        10                        8.25825  
11                        11                        6.69900  
12                        12                        6.83025  
13                        13                        9.47625  
14                        14                        9.37125  
15                        15                        8.28450  

[6 rows x 23 columns]

In [111]:
#selecionando as linhasobservações de indie 1, 5, 10, 15
data.iloc[[1, 5, 10, 15]]

DATA INICIAL  DATA FINAL        REGIÃO    ESTADO           PRODUTO  \
1    2004-05-09  2004-05-15  CENTRO OESTE     GOIAS  ETANOL HIDRATADO   
5    2004-05-09  2004-05-15      NORDESTE     BAHIA  ETANOL HIDRATADO   
10   2004-05-09  2004-05-15      NORDESTE     PIAUI  ETANOL HIDRATADO   
15   2004-05-09  2004-05-15         NORTE  AMAZONAS  ETANOL HIDRATADO   

    NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
1                            387              R$/l                1.162   
5                            408              R$/l                1.383   
10                            89              R$/l                1.573   
15                            53              R$/l                1.578   

    DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
1                   0.114                 0.890  ...   
5                   0.132                 0.999  ...   
10                  0.134                 1.400  ...   
15                  0.123                 1.280  ...   

    PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
1                      0.763                      0.088   
5                      0.957                      0.128   
10                     1.079                      0.078   
15                     1.137                      0.068   

    PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
1                      0.5013                      1.05   
5                      0.5686                      1.35   
10                     0.8857                    1.2499   
15                     0.9515                    1.2109   

   COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO coluna sem nocao  \
1                          0.115   5  2004          DEFAULT   
5                          0.134   5  2004          DEFAULT   
10                         0.072   5  2004          DEFAULT   
15                          0.06   5  2004          DEFAULT   

    coluna a partir de lista  PREÇO MÉDIO REVENDA (dólares)  
1                          1                        6.10050  
5                          5                        7.26075  
10                        10                        8.25825  
15                        15                        8.28450  

[4 rows x 23 columns]

In [112]:
data.iloc[[5, 1, 15, 10]]

DATA INICIAL  DATA FINAL        REGIÃO    ESTADO           PRODUTO  \
5    2004-05-09  2004-05-15      NORDESTE     BAHIA  ETANOL HIDRATADO   
1    2004-05-09  2004-05-15  CENTRO OESTE     GOIAS  ETANOL HIDRATADO   
15   2004-05-09  2004-05-15         NORTE  AMAZONAS  ETANOL HIDRATADO   
10   2004-05-09  2004-05-15      NORDESTE     PIAUI  ETANOL HIDRATADO   

    NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  PREÇO MÉDIO REVENDA  \
5                            408              R$/l                1.383   
1                            387              R$/l                1.162   
15                            53              R$/l                1.578   
10                            89              R$/l                1.573   

    DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
5                   0.132                 0.999  ...   
1                   0.114                 0.890  ...   
15                  0.123                 1.280  ...   
10                  0.134                 1.400  ...   

    PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
5                      0.957                      0.128   
1                      0.763                      0.088   
15                     1.137                      0.068   
10                     1.079                      0.078   

    PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
5                      0.5686                      1.35   
1                      0.5013                      1.05   
15                     0.9515                    1.2109   
10                     0.8857                    1.2499   

   COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO coluna sem nocao  \
5                          0.134   5  2004          DEFAULT   
1                          0.115   5  2004          DEFAULT   
15                          0.06   5  2004          DEFAULT   
10                         0.072   5  2004          DEFAULT   

    coluna a partir de lista  PREÇO MÉDIO REVENDA (dólares)  
5                          5                        7.26075  
1                          1                        6.10050  
15                        15                        8.28450  
10                        10                        8.25825  

[4 rows x 23 columns]

In [113]:
data.iloc[5, 4]

'ETANOL HIDRATADO'

In [114]:
# e assim por diante!

#### **==>  Label-based selection (seleção baseadas em Rótulos)**

`loc`: seleciona elementos do Dataframe, baseado em seus **rótulos** --> row-first, column-second

In [115]:
data.head()

DATA INICIAL  DATA FINAL        REGIÃO              ESTADO  \
0   2004-05-09  2004-05-15  CENTRO OESTE    DISTRITO FEDERAL   
1   2004-05-09  2004-05-15  CENTRO OESTE               GOIAS   
2   2004-05-09  2004-05-15  CENTRO OESTE         MATO GROSSO   
3   2004-05-09  2004-05-15  CENTRO OESTE  MATO GROSSO DO SUL   
4   2004-05-09  2004-05-15      NORDESTE             ALAGOAS   

            PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0  ETANOL HIDRATADO                           127              R$/l   
1  ETANOL HIDRATADO                           387              R$/l   
2  ETANOL HIDRATADO                           192              R$/l   
3  ETANOL HIDRATADO                           162              R$/l   
4  ETANOL HIDRATADO                           103              R$/l   

   PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                1.288                  0.016                  1.19  ...   
1                1.162                  0.114                  0.89  ...   
2                1.389                  0.097                  1.18  ...   
3                1.262                  0.070                  1.09  ...   
4                1.181                  0.078                  1.05  ...   

   PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                     0.825                       0.11   
1                     0.763                      0.088   
2                      0.97                      0.095   
3                      0.83                      0.119   
4                     0.941                      0.077   

   PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                     0.4201                    0.9666   
1                     0.5013                      1.05   
2                     0.5614                     1.161   
3                     0.5991                   1.22242   
4                     0.7441                    1.0317   

  COEF DE VARIAÇÃO DISTRIBUIÇÃO MÊS   ANO coluna sem nocao  \
0                         0.133   5  2004          DEFAULT   
1                         0.115   5  2004          DEFAULT   
2                         0.098   5  2004          DEFAULT   
3                         0.143   5  2004          DEFAULT   
4                         0.082   5  2004          DEFAULT   

   coluna a partir de lista  PREÇO MÉDIO REVENDA (dólares)  
0                         0                        6.76200  
1                         1                        6.10050  
2                         2                        7.29225  
3                         3                        6.62550  
4                         4                        6.20025  

[5 rows x 23 columns]

### 1.5 Selecionando um ou mais atributos (colunas)

In [5]:
#retornando a coluna/atributo (coluna)
type(data['ESTADO'])

pandas.core.series.Series

In [117]:
data.ESTADO #esse acesso só funciona pois nao possui caracteres especiais

0            DISTRITO FEDERAL
1                       GOIAS
2                 MATO GROSSO
3          MATO GROSSO DO SUL
4                     ALAGOAS
                 ...         
106818    RIO GRANDE DO NORTE
106819      RIO GRANDE DO SUL
106820         SANTA CATARINA
106821              SAO PAULO
106822                SERGIPE
Name: ESTADO, Length: 106823, dtype: object

In [118]:
#recuperar todas as linhas mas apenas a coluna
data.loc[:, 'ESTADO']

0            DISTRITO FEDERAL
1                       GOIAS
2                 MATO GROSSO
3          MATO GROSSO DO SUL
4                     ALAGOAS
                 ...         
106818    RIO GRANDE DO NORTE
106819      RIO GRANDE DO SUL
106820         SANTA CATARINA
106821              SAO PAULO
106822                SERGIPE
Name: ESTADO, Length: 106823, dtype: object

In [119]:
#para colunas cujo rotulos possuem caracteres invalidos/especiais
#apenas a seleção via string é valida
data["DATA INICIAL"]

0         2004-05-09
1         2004-05-09
2         2004-05-09
3         2004-05-09
4         2004-05-09
             ...    
106818    2019-06-23
106819    2019-06-23
106820    2019-06-23
106821    2019-06-23
106822    2019-06-23
Name: DATA INICIAL, Length: 106823, dtype: object

In [120]:
#SELECIONANDO MULTIPLAS COLUNAS
data[['PRODUTO', 'ESTADO', 'REGIÃO']]

PRODUTO               ESTADO        REGIÃO
0       ETANOL HIDRATADO     DISTRITO FEDERAL  CENTRO OESTE
1       ETANOL HIDRATADO                GOIAS  CENTRO OESTE
2       ETANOL HIDRATADO          MATO GROSSO  CENTRO OESTE
3       ETANOL HIDRATADO   MATO GROSSO DO SUL  CENTRO OESTE
4       ETANOL HIDRATADO              ALAGOAS      NORDESTE
...                  ...                  ...           ...
106818               GNV  RIO GRANDE DO NORTE      NORDESTE
106819               GNV    RIO GRANDE DO SUL           SUL
106820               GNV       SANTA CATARINA           SUL
106821               GNV            SAO PAULO       SUDESTE
106822               GNV              SERGIPE      NORDESTE

[106823 rows x 3 columns]

Como o rótulo da coluna 'DATA INICIAL' **possui espaço**, não é possível acessá-la como método: `data.DATA INICIAL`

### 1.6 Removendo um Atributo (Coluna) do Data Frame

Como vimos anteriormente, o atributo 'Unnamed: 0' parece ser um **ruído** em nosso dataset. Desta maneira, vamos eliminá-lo,

In [121]:
data

DATA INICIAL  DATA FINAL        REGIÃO               ESTADO  \
0        2004-05-09  2004-05-15  CENTRO OESTE     DISTRITO FEDERAL   
1        2004-05-09  2004-05-15  CENTRO OESTE                GOIAS   
2        2004-05-09  2004-05-15  CENTRO OESTE          MATO GROSSO   
3        2004-05-09  2004-05-15  CENTRO OESTE   MATO GROSSO DO SUL   
4        2004-05-09  2004-05-15      NORDESTE              ALAGOAS   
...             ...         ...           ...                  ...   
106818   2019-06-23  2019-06-29      NORDESTE  RIO GRANDE DO NORTE   
106819   2019-06-23  2019-06-29           SUL    RIO GRANDE DO SUL   
106820   2019-06-23  2019-06-29           SUL       SANTA CATARINA   
106821   2019-06-23  2019-06-29       SUDESTE            SAO PAULO   
106822   2019-06-23  2019-06-29      NORDESTE              SERGIPE   

                 PRODUTO  NÚMERO DE POSTOS PESQUISADOS UNIDADE DE MEDIDA  \
0       ETANOL HIDRATADO                           127              R$/l   
1       ETANOL HIDRATADO                           387              R$/l   
2       ETANOL HIDRATADO                           192              R$/l   
3       ETANOL HIDRATADO                           162              R$/l   
4       ETANOL HIDRATADO                           103              R$/l   
...                  ...                           ...               ...   
106818               GNV                             7             R$/m3   
106819               GNV                            23             R$/m3   
106820               GNV                            24             R$/m3   
106821               GNV                            52             R$/m3   
106822               GNV                             4             R$/m3   

        PREÇO MÉDIO REVENDA  DESVIO PADRÃO REVENDA  PREÇO MÍNIMO REVENDA  ...  \
0                     1.288                  0.016                 1.190  ...   
1                     1.162                  0.114                 0.890  ...   
2                     1.389                  0.097                 1.180  ...   
3                     1.262                  0.070                 1.090  ...   
4                     1.181                  0.078                 1.050  ...   
...                     ...                    ...                   ...  ...   
106818                3.574                  0.065                 3.499  ...   
106819                3.401                  0.129                 3.230  ...   
106820                2.912                  0.190                 2.699  ...   
106821                3.020                  0.229                 2.699  ...   
106822                3.697                  0.005                 3.689  ...   

        PREÇO MÉDIO DISTRIBUIÇÃO DESVIO PADRÃO DISTRIBUIÇÃO  \
0                          0.825                       0.11   
1                          0.763                      0.088   
2                           0.97                      0.095   
3                           0.83                      0.119   
4                          0.941                      0.077   
...                          ...                        ...   
106818                     2.756                          0   
106819                     2.451                      0.402   
106820                     1.998                          0   
106821                     2.374                      0.165   
106822                     2.518                          0   

        PREÇO MÍNIMO DISTRIBUIÇÃO PREÇO MÁXIMO DISTRIBUIÇÃO  \
0                          0.4201                    0.9666   
1                          0.5013                      1.05   
2                          0.5614                     1.161   
3                          0.5991                   1.22242   
4                          0.7441                    1.0317   
...                           ...                       ...   
106818                      2.756                     2.756   
106819             

In [122]:
del data['coluna sem nocao']
del data['coluna a partir de lista']
del data['PREÇO MÉDIO REVENDA (dólares)']

In [123]:
del data['Unnamed: 0']

KeyError: 'Unnamed: 0'

### 1.7 Salvando um Data Frame

Para salvarmos um Data Frame para um **arquivo CSV**, basta usarmos o método `.to_csv`. <br/>
Por padrão, esse método **salva os índices da tabela como uma coluna no CSV**.<br/>
Como no geral tais índices são números de 0 a n-1, não há necessidade para isso (veja que removemos anteriormente a coluna 'Unnamed: 0' que foi justamente esse caso).<br/>
Desta forma, utilize o parâmetro: `index=False`.

Por padrão, o método utilizará a ',' como separador das colunas. Caso queira alterá-lo, utilize o parâmetro `sep`.

In [ ]:
data.to_csv('./datasets/GasPricesinBrazil_2004-2019.csv', index=False) 
#esse index false diz para nao criar outro id no dataframe

### 1.8 Seleção Condicional: Filtrando amostras

Durante nossas análise exploratórias, frequentemente filtraremos nossas amostras, a partir de certas **condições**, para fins de análise mais específica. <br/>
Existem algumas maneiras de fazermos tal filtragem. Antes disso, vamos carregar nosso dataset pré-processado que salvamos no item anterior.

In [ ]:
data = pd.read_csv('./datasets/GasPricesinBrazil_2004-2019.csv')
data

In [ ]:
#unique() mostra os valores unicos de uma Series
data['ESTADO'].unique()

#### **Selecionando apenas os preços dos Postos de São Paulo**

##### **==> Alternativa 1: Seleção Condicional (Comparações diretas)**

O código abaixo retorna uma ***Series ('array') de booleans***, com o número de linhas (amostras) do Data Frame, que informa os registros de preços dos postos do _estado de São Paulo_ (True).

In [ ]:
data['ESTADO'] == 'SAO PAULO'
#faz uma comparação elemento por elemento retornando uma Series de booleans

In [ ]:
selecao = data['ESTADO'] == 'SAO PAULO'
selecao

In [ ]:
#usando a 'selecao' com mascara para visualização
data[selecao] #ela retorna só os valores que deram true la em cima (sao paulo)

In [ ]:
data.loc[selecao]

In [ ]:
prm = data['PREÇO MÁXIMO REVENDA'] >= 1.150
prm #podemos fazer assim tbm, passando == ou >= ou <= etc... Sempre retornara uma mascara para ser atribuida

In [ ]:
data[prm]

Para **filtrarmos** os registros de postos do estado de São Paulo:

In [ ]:
data[selecao]

O resultado é um Data Frame com _apenas_ os registros desejados após a **filtragem**.<br/>
Podemos ainda utilizar o método `loc` para o mesmo fim:

##### **==> Alternativa 2: Utilizando o método `query`**

`query` filtra linhas de um DataFrame baseado em uma **query (pergunta)**.

In [ ]:
data.query('ESTADO == "SAO PAULO"')


<br/>

Uma boa prática é **salvar o Data Frame filtrado em uma nova variável**. Isso simplifica a complexidade do código para futuras análise feita para os postos de São Paulo.

In [ ]:
postos_sp =data.query('ESTADO == "SAO PAULO"')

In [ ]:
postos_sp

In [ ]:
postos_sp.reset_index()
#resetar o index para ficar mais 'certinho' a sequencia numerica
#note que é gerada uma coluna index relacionada aos antigos indices... veremos abaixo + sobre isso

In [ ]:
postos_sp.reset_index(drop=True, inplace=True)
postos_sp
#este comando exclui aquele indice que ficou gravado, mantendo apenas o novo, organizado
#entretantooo, temos que colocar o inplace pq ele nao muda, só gera uma cópia, o inplace garante a mudança!

In [ ]:
#alternativa comum que nao precisa do inplace
postos_sp = data.query('ESTADO == "SAO PAULO"').reset_index(drop=True)
postos_sp

#### **Selecionando registros de postos do Rio de Janeiro com Preços acima de 2 reais**

In [ ]:
data.head()

In [ ]:
postos_rj = (data['ESTADO'] == 'RIO DE JANEIRO') & (data['PREÇO MÁXIMO REVENDA'] > 2.0)
data[postos_rj]
#note os parenteses para cada um

Note que o resultado da seleção continua sendo uma _Series de booleans_ com o _mesmo número de linhas/amostras do DataFrame_, de modo que cada linha possuirá um valor booleano indicando se o posto é do Rio de Janeiro e o preço aferido do combustível é maior que 2 reais (True) ou não.

O símbolo **&** significa **AND** na comparação. Essa nomenclatura do python/pandas é diferente das nomenclaturas tradicionais (&&). <br/>
Similarmente:
- **|** representa o **OR** (não é ||)
- **~** representa o **NOT** (não é !)

In [ ]:
selecao = (data['PRODUTO'] == 'GLP') & (data['ANO'] >= 2019) & (data['PREÇO MÁXIMO REVENDA'] > 89.00)
data[selecao]

Alternativamente, poderíamos usar o método `query` para fazermos tal seleção. Porém, isso não é possível especificamente para esse caso, pois o rótulo da coluna 'PREÇO MÉDIO REVENDA' possui caracteres inválidos para o método (cedilha, acentos) 

In [ ]:
# Não funciona por conta do acento no médio, sem ele funcionaria!!!
# data.query('ESTADO=="RIO DE JANEIRO" and PREÇO MÉDIO REVENDA > 2')
#repare que para query, se usa and ao invés de &

In [ ]:
rj_sp = data.query('ESTADO == "SAO PAULO" or ESTADO == "RIO DE JANEIRO"')
rj_sp

**Aprofundando mais ainda**

A primeira comparação `(data['ESTADO'] == 'RIO DE JANEIRO')` checa, linha a linha (amostra a amostra) do DataFrame, quais são aquelas cujo o estado é RIO DE JANEIRO. Nenhuma averiguação de preços é feita nesse momento. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

A segunda comparação `(data['PREÇO MÉDIO REVENDA'] > 2)` checa, linha a linha (amostra a amostra) do DataFrame, quais são os registro cujo preço do combustível é maior que 2 reais. Note que essa comparação checará os postos de **TODOS os estados**. Como resultado, temos uma Series de booleans que responde **apenas** a essa "pergunta" feita.

Por fim, as duas "perguntas" são unidas pelo AND (&) que retorna a "pergunta completa" que fizemos.

Alguns podem argumentar que tal abordagem é **ineficiente**, uma vez que, para cada condição ("pergunta"), estamos varrendo todas as linhas do DataFrame. <br/>
O Pandas _tenta otimizar_ isso ao máximo por de trás dos panos. Mas, de fato, de tivermos um dataset **muito grande** (centenas de milhares de linhas), tal abordagem se tornará _lenta_.

Assim, poderíamos fazer filtragem com múltiplos condicionais em partes:

In [ ]:
selecao1 = data['ESTADO'] == 'RIO DE JANEIRO'
postos_rj = data[selecao1]
postos_rj

In [ ]:
postos_rj2 = postos_rj['PREÇO MÉDIO REVENDA'] > 2
postos_rj2

In [ ]:
postos_rj_com_valores_filtrados = postos_rj[postos_rj2]
postos_rj_com_valores_filtrados

#### **Selecionando registros de postos de São Paulo ou do Rio de Janeiro com Gasolina Comum acima de 2 reais**

Podemos fazer a solução do "jeito mais lento", percorrendo o DataFrame inteiro _múltiplas vezes_:

In [ ]:
sp_rj_postos = (data['ESTADO'] == 'SAO PAULO') | (data['ESTADO'] == 'RIO DE JANEIRO')
sp_rj_postos = data[sp_rj_postos]
sp_rj_postos

In [ ]:
postos_filtrados = (sp_rj_postos['PRODUTO'] == 'GASOLINA COMUM') & (sp_rj_postos['PREÇO MÉDIO REVENDA'] > 2.0)
postos_filtrados = sp_rj_postos[postos_filtrados]
postos_filtrados

<br/>

Alternativamente:

#### **Selecionando registros dos anos de 2008, 2010 e 2012**

In [ ]:
data['ANO'].unique()

**ALTERNATIVA 1**

In [ ]:
selecao_anual = (data['ANO'] == 2008) | (data['ANO'] == 2010) | (data['ANO'] == 2012)
data[selecao_anual]

**ALTERNATIVA 2**

In [ ]:
lista_de_anos = [2008, 2010, 2012]

In [ ]:
selecao = data['ANO'].isin(lista_de_anos)
data[selecao]

<br>
<h3> Alternativa 3 </h3>

In [ ]:
lista_de_anos

In [ ]:
#utilizando query
data.query('ANO in @lista_de_anos') # o @ indica que é para consultar uma variavel existente!

### **Iterando com DataFrames**

#### For-each `DataFrame.iterrows()` (LENTO ==> apenas indicado para iterar pequenos conjunto de dados)

In [6]:
for index, row in data.head(10).iterrows(): #pra cada indice e linha dentro do dataset sera iterado ...
    print(f'indice {index} ==> {row["ESTADO"]}') 

indice 0 ==> DISTRITO FEDERAL
indice 1 ==> GOIAS
indice 2 ==> MATO GROSSO
indice 3 ==> MATO GROSSO DO SUL
indice 4 ==> ALAGOAS
indice 5 ==> BAHIA
indice 6 ==> CEARA
indice 7 ==> MARANHAO
indice 8 ==> PARAIBA
indice 9 ==> PERNAMBUCO


<h2>2. Preparação dos dados</h2>
<hr/>

### 2.1 Removendo amostras com valores vazios (null / nan) no dataset

In [ ]:
data_pre = data.copy()

In [ ]:
data_pre.info()

De um total de 106823 observações, **não há valores null** / nan para nenhum atributo. Mas, veremos que não é bem assim neste caso específico.<br/><br/>

### 2.2 Conversão de tipos de atributos

O pandas automaticamente reconhece os tipos de dados de cada coluna. <br/>
Porém, existem alguns atributos que estão com seus tipos errados: P. ex., "PREÇO MÉDIO DISTRIBUIÇÃO" deveria ser ```float64``` e não ```object```.<br/>
Nestes casos, muito provavelmente algumas amostras têm um string ao invés de um número para tais atributos. <br/>

Os atributos *"DATA INICIAL"* e *"DATA FINAL"* deveriam ser do tipo `datetime`.

Em outros casos, alguns **atributos categóricos** são ```objects```, mas poderiam ter o tipo ```category```, que é um tipo especial do pandas. <br/>
Este tipo é necessário para se utilizar algumas funções específicas do pandas. <br/>
**Não** converteremos para este tipo por ora.

#### **Datas**
Como os atributos de data do datset já estão em um formato de data aceitável (YYYY-MM-DD), não precisamos forçar nenhuma conversão nesse sentido.

In [ ]:
data_pre['DATA INICIAL'] = pd.to_datetime(data_pre['DATA INICIAL'])
data_pre['DATA FINAL'] = pd.to_datetime(data_pre['DATA FINAL'])
data_pre.info()

#### **Dados Numéricos**

In [ ]:
data_pre

In [ ]:
for atributo in ['MARGEM MÉDIA REVENDA', 'PREÇO MÉDIO DISTRIBUIÇÃO', 'DESVIO PADRÃO DISTRIBUIÇÃO', 'PREÇO MÍNIMO DISTRIBUIÇÃO', 'PREÇO MÁXIMO DISTRIBUIÇÃO', 'COEF DE VARIAÇÃO DISTRIBUIÇÃO']:
    data_pre[atributo] = pd.to_numeric(data_pre[atributo], errors='coerce')
    #em caso de erro na conversão (p. ex., uma string que nao representa um numero), um valor vazio NULL / NaN
   #será atribuido no lugar

In [ ]:
data_pre.info()
#observe que o numero de Non-NULL mudou.
#muito provavelmente os numeros se tornaram NaN por algum erro na transformação.

<br/>

Note que temos vários valores ***null*** agora **após a *conversão de tipos***. Vamos checá-los com mais cuidado nos dados originais e preprocessados.

### 2.3 Limpeza de dados

In [ ]:
eh_null = data_pre['PREÇO MÉDIO DISTRIBUIÇÃO'].isnull()
eh_null

In [ ]:
data[eh_null]
#se compararmos o DF original com nossa mascara, poderemos ver o que foi passado de fato e entender...
#o porque a transformaçao de str (por exemplo) falhou quando tentamos passar para int.

Várias amostras possuem a _string '-'_ em algumas colunas ao invés de um número de fato. Ou seja, não há aferições destes atributos para estas amostras. <br/>

<br/>

Poderíamos **preencher os valores NaN com um valor padrão**. Para isso, basta usar o método `.fillna`.

In [ ]:
#retorna uma cópia do 'data_pré' com todos os valores NaN de todas as colunas agora preenchidos com 0
data_pre_fill = data_pre.fillna(0)
data_pre_fill

In [ ]:
data_pre_fill[eh_null] #nao alteramos o DF preprocessado (original)

In [ ]:
#retorna uma cópia do DF com todos os valores NaN das colunas citadas.
data_pre_fill = data_pre.fillna(value={
    'PREÇO MÉDIO DISTRIBUIÇÃO':10,
    'DESVIO PADRÃO DISTRIBUIÇÃO':20,
    'PREÇO MÍNIMO DISTRIBUIÇÃO':30,
    'PREÇO MÁXIMO DISTRIBUIÇÃO':'vazio'
})
data_pre_fill

In [ ]:
data_pre_fill[eh_null]

<br/>

Por mais que a função `fillna` seja interessante e útil em muitos casos, no problema em questão estamos interessados em analisar precisamente, p. ex., o **'PREÇO MÉDIO DISTRIBUIÇÃO'**.<br/>
A fim de não termos valores _sintéticos_ gerados pelo `fillna`, que possam atrapalhar nossa análise, iremos **remover (drop) todas as amostras que possuem qualquer valor NaN** para quaisquer atributos/colunas. <br/>

Para isso, basta utilizarmos o método `dropna`.

In [ ]:
#remove no próprio dataframe, todas as linhas/registros com valores NaN em quaisquer colunas/atributos
data_pre.dropna(inplace=True)
data_pre.info()

Nosso data frame agora, após essa _limpeza_, ficou mais enxuto, contentdo 103392 registros frente aos 106823 registros originais. <br/>

Essas são apenas algumas das possíveis _técnicas de limpeza de dados_. Outras estratégias, p. ex., **confiam na detecção de outliers**, que não veremos neste curso.

#### **Salvando o Dataset Preprocessado**

In [ ]:
data_pre.to_csv('./datasets/GasPricesinBrazil_2004-2019_preprocessado.csv', index=False)

<h2>3. Estatísticas Descritivas</h2>
<hr/>

O Pandas fornecem algumas funções/métodos ue computam certas estatísticas descritivas.

`describe`: exibe várias **estatísticas descritivas** para os _atributos_ de um dataframe ou para uma _Series_.

In [ ]:
data_final = pd.read_csv('./datasets/GasPricesinBrazil_2004-2019_preprocessado.csv')

In [ ]:
data_final

<br/>

Como o resultado do `describe` de um _dataframe_ é outro _dataframe_, podemos filtrar apenas algumas colunas.

In [ ]:
data_final.describe()

In [ ]:
data_final['PREÇO MÉDIO REVENDA'].describe()

**Acessando apenas algumas estatísticas**

In [ ]:
stats = data_final.describe()

In [ ]:
stats[['PREÇO MÉDIO REVENDA', 'PREÇO MÍNIMO REVENDA', 'PREÇO MÁXIMO REVENDA']]

In [ ]:
data_final[['PREÇO MÉDIO REVENDA', 'PREÇO MÍNIMO REVENDA', 'PREÇO MÁXIMO REVENDA']].describe()

<br/>

`mean`, `std`, `min`, etc: cada uma das estatísticas do `describe` podem ser computadas individualmente:

In [ ]:
stats

In [ ]:
stats.loc[['min', 'max', 'mean']]

#### Qual é o menor preço mínimo de revenda?

In [ ]:
stats['PREÇO MÍNIMO REVENDA'].loc['min']

#### Qual é a média e desvio padrão dos preços mínimos de revenda?

In [ ]:
stats['PREÇO MÍNIMO REVENDA'].loc['std']

#### Quais são os estados considerados?

In [ ]:
data_final['ESTADO'].unique()

#### Quantos registros (aferições) cada estado possui?

`.value_counts()`:  Conta a frequência dos valores de uma dada variável (de preferência, _categórica_).

In [ ]:
data_final['ESTADO'].value_counts()

<h2>4. Executando funções para cada item de um DataFrame ou Series</h2>
<hr/>

Uma alternativa ao `for-loop` que vimos anteriormente e que é _lento_, é usarmos _funções próprias do pandas_ que **aplicam/mapeiam uma dada função a todos os elementos de um DataFrame ou Series**, retornando novos elementos "transformados".


<img src='./imagens/apply_map_applymap.png' width=300/>


Fonte: https://towardsdatascience.com/introduction-to-pandas-apply-applymap-and-map-5d3e044e93ff

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4], 
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]}, 
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])

In [ ]:
df

`apply()`: usado para aplicar uma função ao longo de um eixo de um DataFrame ou em valores de uma Series.

<img src='./imagens/pandas_axis.jpg' width=500/>

Fonte: https://www.allthesnippets.com/browse/pandas/df_axis.html

In [128]:
def nossa_soma(series):
    return series.sum() #retorna a soma de todos os valores de  uma linha

In [129]:
df['SOMA(A, B, C)'] = df.apply(nossa_soma, axis=1)
df

A   B    C  SOMA(A, B, C)
Linha 1  1  10  100          111.0
Linha 2  2  20  200          222.0
Linha 3  3  30  300          333.0
Linha 4  4  40  400          444.0

<img src='./imagens/apply_axis_1.png' width=250/>

In [131]:
df.loc['linha 5'] = df.apply(nossa_soma, axis=0)
df

A      B       C  SOMA(A, B, C)
Linha 1   1.0   10.0   100.0          111.0
Linha 2   2.0   20.0   200.0          222.0
Linha 3   3.0   30.0   300.0          333.0
Linha 4   4.0   40.0   400.0          444.0
linha 5  10.0  100.0  1000.0         1110.0

<img src='./imagens/apply_axis_0.png' width=250/>

##### Usando `lambda` functions

In [134]:
df['MEDIA(A,B,C)'] = df[['A','B','C']].apply(lambda series: series.mean(), axis=1)

Linha 1     37.0
Linha 2     74.0
Linha 3    111.0
Linha 4    148.0
linha 5    370.0
dtype: float64

<img src='./imagens/apply_axis_1_mean.png' width=350/>

<br/>

`applymap()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de um DataFrame.

In [ ]:
df = pd.DataFrame({ 'A': [1, 2, 3, 4], 
                    'B': [10, 20, 30, 40],
                    'C': [100, 200, 300, 400]}, 
                     index=['Linha 1', 'Linha 2', 'Linha 3', 'Linha 4'])
df

<br/>

`map()`: usado para aplicar uma função para **cada elemento** (_element-wise_) de uma _Series_.

<h2>5. Agrupamento</h2>
<hr/>

`groupby`: Usado para criar **grupo de elementos** (e.x., baseado nos valores de um atributo). <br/>
**Funções** podem então ser aplicadas para os _elementos de cada grupo_, de modo que os **resultados de cada grupo são combinados**.

<br/>

Também podemos ter agrupamentos por mais de um atributo.

<br/>

`.agg`: **agrega (roda)** uma série de funções para os elementos de um dataframe ou de grupos de um dataframe.

In [ ]:
df = pd.DataFrame([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9],
                   [np.nan, np.nan, np.nan]],
                  columns=['A', 'B', 'C'])
df

<h2>6. Ordenação</h2>
<hr/>

In [ ]:
notas = pd.DataFrame({
    'nome': ['João', 'Maria', 'José', 'Alice'],
    'idade': [20, 21, 19, 20],
    'nota_final': [5.0, 10.0, 6.0, 10.0]
})
notas

`.sort_values()`: ordena valores ao longo de um eixo.

Por padrão, o método retorna uma cópia dos dados ordenados em **ordem crescente (ascendente)**. Podemos alterar isso pelo argumento `ascending`.

<br/>

Podemos ordenar a partir de **mais de uma coluna**:

Ordena os registros, primeiramente, pela coluna 'nota_final' em **ordem descrente**. <br/>
Então, reordena os registros _"empatados"_, ou seja, com a **mesma nota final**, em _ordem alfabética_ (ordem crescente).

<br/>

Note que o dataframe original **não foi alterado** após a ordenação.

Para alterá-lo, use o argumento `inplace=True`:

<h2>7. Exercícios</h2>
<hr/>

Vamos aplicar os conceitos que vimos em alguns exercícios. <br/>
Para isso, utilizaremos o dataset de _preços de combustíveis no Brasil_.

Como há apenas medições de janeiro a junho para o ano de 2019, resolvemos **remover os dados** deste ano da análise.

Note que temos um novo dataframe com 99739 linhas, mas com índices fora desse intervalo. <br/>
Acontece que os registros mantiveram seus índices originais após a query. <br/>

Para resetar os índices de _0 a num_linhas-1_, basta usarmos o método `.reset_index()`. 

Os índices agora foram **resetados**. Porém, os índices antigos se transformaram em _uma nova coluna_ chamada 'index'. <br/>
Para removê-la durante o _reset_, basta passarmos o argumento `drop=True`.

### 7.1 Qual a proporção de postos pesquisados para cada combustível em cada região

### 7.2 Como os preços da Gasolina Comum em São Paulo variaram em 2018?

#### **Estatísticas Descritivas**

### 7.3 Como os preços da Gasolina Comum e do Etanol em São Paulo variaram em 2018?

<h2>8. Assuntos para continuar os estudos</h2>
<hr/>

- join
- concat
- plot
- data cleaning